### PR TITLE
feature/PU-137 Predictions update

### PR DESCRIPTION
**Pull request by oleg1std**
Adds a service to publish predictions to the Redis Stream

### CODE REVIEW

File: HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿namespace HappyTravel.Nakijin.Api.Models.PredictionsUpdate
+{
+    public enum EntryTypes
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs:3]: Ass I remember we decided to not name items of enum by verbs
And maybe EventType will be more understandable?

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            _stream = updateOptions.Value.Stream;
+            _logger = logger;
+            _database = redisCacheClient.GetDbFromConfiguration().Database;
+            if (!InitStream())
+                ClearStream();
+        }
+        
+        
+        public async Task Publish(List<Location> locations, EntryTypes type, CancellationToken cancellationToken = default)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs:31]: I think we can have here Publish method with one location

File: HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs (hunk 1 above)
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs:3]: fixed

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs (hunk 2 above)
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs]: the method has been added

**Review by kirillta — CHANGES_REQUESTED** [review]

File: HappyTravel.Nakijin.Api/Infrastructure/DefaultSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Text.Json;
+using StackExchange.Redis.Extensions.Core;
+
+namespace HappyTravel.Nakijin.Api.Infrastructure
+{
+    public class DefaultSerializer: ISerializer
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Infrastructure/DefaultSerializer.cs:6]: fixed

File: HappyTravel.Nakijin.Api/Infrastructure/PredictionUpdateOptions.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace HappyTravel.Nakijin.Api.Infrastructure
+{
+    public class PredictionUpdateOptions
+    {
+        public string Stream { get; set; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Infrastructure/PredictionUpdateOptions.cs:5]: fixed

File: HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            
+            services.AddStackExchangeRedisExtensions<DefaultSerializer>(s 
+                => new ()
+                {
+                    Hosts = new []
+                    {
+                        new RedisHost
+                        {
+                            Host = redisEndpoint,
+                            Port = 6379
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs:47]: fixed

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        
+        
+        NameValueEntry[] Build(List<Location> batchOfLocations, EventTypes type) 
+            => batchOfLocations.Select(l => Build(l, type))
+                .ToArray();
+
+
+        private readonly AsyncLock _mutex = new ();
+        private readonly string _stream;
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs:96]: fixed

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            var streamInfo = _database.StreamInfo(_stream);
+            if (!streamInfo.FirstEntry.IsNull && !streamInfo.LastEntry.IsNull)
+            {
+                var messagesToClear = _database.StreamRange(_stream, streamInfo.FirstEntry.Id, streamInfo.LastEntry.Id);
+                _database.StreamDelete(_stream, messagesToClear.Select(m => m.Id).ToArray());
+            }
+        }
+
+        
+        NameValueEntry Build(Location location, EventTypes type)
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs:83]: fixed

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                foreach (var batchOfLocations in ListHelper.Split(locations, batchSize))
+                {
+                    await _database.StreamAddAsync(_stream, Build(batchOfLocations, type));
+                    _logger.LogLocationsPublished( $"{batchOfLocations} locations have been published");
+                }
+            }
+        }
+
+        
+        private bool InitStream()
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs:52]: InitStreamIfNeeded ?

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            _logger.LogLocationsPublished( $"Location '{location.HtId}' has been published");
+        }
+
+        
+        public async Task Publish(List<Location> locations, EventTypes type, CancellationToken cancellationToken = default)
+        {
+            const int batchSize = 1000;
+            using (await _mutex.LockAsync(cancellationToken))
+            {
+                foreach (var batchOfLocations in ListHelper.Split(locations, batchSize))
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs:43]: fixed

File: HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            var role = configuration["PredictionsUpdate:Role"];
+            using var vaultClient = new VaultClient.VaultClient(new VaultOptions
+            {
+                BaseUrl = new Uri(EnvironmentVariableHelper.Get("Vault:Endpoint", configuration)),
+                Engine = configuration["Vault:Engine"],
+                Role = role
+            });
+            var token = configuration[configuration["Vault:Token"]];
+            vaultClient.Login(token).GetAwaiter().GetResult();
+            var redisOptions = vaultClient.Get(configuration["PredictionsUpdate:Redis"]).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (kirillta) [reply HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs]: This one still not fixed

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/IPredictionsUpdateService.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using HappyTravel.Nakijin.Api.Models.LocationServiceInfo;
+using HappyTravel.Nakijin.Api.Models.PredictionsUpdate;
+
+namespace HappyTravel.Nakijin.Api.Services.PredictionsUpdate
+{
+    public interface IPredictionsUpdateService
+    {
+        Task Publish(Location location, EventTypes type);
```
[the remark this answers is not in the document]
Comment (kirillta) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/IPredictionsUpdateService.cs]: And this one

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kirillta) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs]: Probably, yes

**Review by kirillta — COMMENTED** [review]
Some issues was missed due other changes

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/PredictionsUpdateService.cs]: fixed

File: HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+﻿namespace HappyTravel.Nakijin.Api.Models.PredictionsUpdate
+{
+    public enum EventTypes
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs]: fixed

File: HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+    {
+        Add = 0,
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Models/PredictionsUpdate/EntryType.cs]: fixed

File: HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using StackExchange.Redis.Extensions.Core.Configuration;
+
+namespace HappyTravel.Nakijin.Api.Infrastructure
+{
+    public static class PredictionsUpdateExtensions
+    {
+        public static IServiceCollection AddPredictionsUpdate(this IServiceCollection services, IConfiguration configuration, IWebHostEnvironment environment)
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs]: fixed

File: HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Infrastructure/PredictionsUpdateExtensions.cs]: fixed

File: HappyTravel.Nakijin.Api/Services/PredictionsUpdate/IPredictionsUpdateService.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/PredictionsUpdate/IPredictionsUpdateService.cs]: The service has been renamed

File: HappyTravel.Nakijin.Api/Models/PredictionsUpdate/LocationEntry.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+﻿using HappyTravel.Nakijin.Api.Models.LocationServiceInfo;
+
+namespace HappyTravel.Nakijin.Api.Models.PredictionsUpdate
+{
+    public readonly struct LocationEntry
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Models/PredictionsUpdate/LocationEntry.cs]: Struct has been renamed

**Review by kirillta — APPROVED** [review]

**Pull request merged**